File: applayer/ruleinfo.c
/*
 * Copyright (c) 2007, Novell Inc.
 *
 * This program is licensed under the BSD license, read LICENSE.BSD
 * for further information
 */

/*
 * Ruleinfo
 *
 * Information about a single solver rule.
 *
 */

#include <stdlib.h>
#include "job.h"

#include "ruleinfo.h"

#include "applayer.h"
#include "solverdebug.h"

Ruleinfo *
ruleinfo_new( Solver *solver, Id rule )
{
  Ruleinfo *ri = (Ruleinfo *)calloc( 1, sizeof( Ruleinfo ));
  ri->solver = solver;
  ri->id = rule;
  ri->cmd = solver_ruleinfo((Solver *)solver, rule, &(ri->source), &(ri->target), &(ri->dep));
  return ri;
}


char *
ruleinfo_string( const Ruleinfo *ri )
{
  app_debugstart(ri->solver->pool,SAT_DEBUG_RESULT);
  solver_printproblemruleinfo(ri->solver, ri->id);
  return app_debugend();
}


void
ruleinfo_free( Ruleinfo *ri )
{
  free( ri );
}


const char *
ruleinfo_command_string(const Ruleinfo *ri)
{
  switch (ri->cmd) 
    {
#define rulecase(r) case r: return #r; break
      rulecase(SOLVER_RULE_UNKNOWN);
      rulecase(SOLVER_RULE_RPM);
      rulecase(SOLVER_RULE_RPM_NOT_INSTALLABLE);
      rulecase(SOLVER_RULE_RPM_NOTHING_PROVIDES_DEP);
      rulecase(SOLVER_RULE_RPM_PACKAGE_REQUIRES);
      rulecase(SOLVER_RULE_RPM_SELF_CONFLICT);
      rulecase(SOLVER_RULE_RPM_PACKAGE_CONFLICT);
      rulecase(SOLVER_RULE_RPM_SAME_NAME);
      rulecase(SOLVER_RULE_RPM_PACKAGE_OBSOLETES);
      rulecase(SOLVER_RULE_RPM_IMPLICIT_OBSOLETES);
      rulecase(SOLVER_RULE_UPDATE);
      rulecase(SOLVER_RULE_FEATURE);
      rulecase(SOLVER_RULE_JOB);
      rulecase(SOLVER_RULE_JOB_NOTHING_PROVIDES_DEP);
      rulecase(SOLVER_RULE_DISTUPGRADE);
      rulecase(SOLVER_RULE_INFARCH);
      rulecase(SOLVER_RULE_LEARNT);
#undef rulecase
    default:
      break;
    }
  return "Unknown";
}


int
ruleinfo_command(const Ruleinfo *ri)
{
  return ri->cmd;
}

/*
 * If cmd == SOLVER_RULE_JOB
 * then source -> job id
 *      target -> job cmd
 *      dep    -> job data (depending on job cmd !)
 */

Job *
ruleinfo_job(const Ruleinfo *ri)
{
  if (ri->cmd != SOLVER_RULE_JOB)
    return NULL;
  return job_new( ri->solver->pool, ri->target, ri->source);
}


XSolvable *
ruleinfo_source(const Ruleinfo *ri)
{
  if (ri->cmd == SOLVER_RULE_JOB)
    return NULL;
  return ri->source ? xsolvable_new( ri->solver->pool, ri->source ) : NULL;
}


XSolvable *
ruleinfo_target(const Ruleinfo *ri)
{
  if (ri->cmd == SOLVER_RULE_JOB)
    return NULL;
  return ri->target ? xsolvable_new( ri->solver->pool, ri->target ) : NULL;
}


Relation *
ruleinfo_relation(const Ruleinfo *ri)
{
  if (ri->cmd == SOLVER_RULE_JOB)
    return NULL;
  return ri->dep ? relation_new( ri->solver->pool, ri->dep ) : NULL;
}

File: bindings/ruby/satsolver/ruleinfo.rb
#
# satsolver/ruleinfo
#

module Satsolver
  class Ruleinfo
    def to_s
      case command
      when SOLVER_RULE_DISTUPGRADE:
	"%s does not belong to a distupgrade repository" % source
      when SOLVER_RULE_INFARCH:
	"%s has inferior architecture" % source
      when SOLVER_RULE_UPDATE:
	"problem with installed package %s" % source
      when SOLVER_RULE_JOB:
	"job request"
      when SOLVER_RULE_JOB_NOTHING_PROVIDES_DEP:
	"nothing provides requested %s" % relation
      when SOLVER_RULE_RPM:
	"some dependency problem"
      when SOLVER_RULE_RPM_NOT_INSTALLABLE:
	"package %s is not installable" % source
      when SOLVER_RULE_RPM_NOTHING_PROVIDES_DEP:
	"nothing provides %s needed by %s" % [ relation, source ]
      when SOLVER_RULE_RPM_SAME_NAME:
	"%s updates %s" % [ source, target ]
      when SOLVER_RULE_RPM_PACKAGE_CONFLICT:
	"package %s conflicts with %s provided by %s" % [ source, relation, target ]
      when SOLVER_RULE_RPM_PACKAGE_OBSOLETES:
	"package %s obsoletes %s provided by %s" % [ source, relation, target ]
      when SOLVER_RULE_RPM_IMPLICIT_OBSOLETES:
	"package %s implicitely obsoletes %s provided by %s" % [ source, relation, target ]
      when SOLVER_RULE_RPM_PACKAGE_REQUIRES:
	"package %s requires %s" % [ source, relation ]
      when SOLVER_RULE_RPM_SELF_CONFLICT:
	"package %s conflicts with %s provided by itself" % [ source, relation ]
      when SOLVER_RULE_UNKNOWN:
        "unknown rule"
      when SOLVER_RULE_FEATURE:
        "feature rule"
      when SOLVER_RULE_LEARNT:
	"learnt rule"
      else
	"***BAD RULE [%d]***" % command
      end
    end
  end
end
